pkg/plugins/policies/core/rules/merge.go

```diff
@@ -10,57 +10,71 @@ import (
 	jsonpatch "github.com/evanphx/json-patch/v5"
 
 	core_model "github.com/kumahq/kuma/pkg/core/resources/model"
+	"github.com/kumahq/kuma/pkg/util/pointer"
 )
 
 // indicates that all slices that start with this prefix will be appended, not replaced
 const appendSlicesPrefix = "Append"
 
-func MergeConfs(confs []interface{}) (interface{}, error) {
+func MergeConfs(confs []interface{}) ([]interface{}, error) {
 	if len(confs) == 0 {
 		return nil, nil
 	}
 
-	resultBytes := []byte{}
-	for _, conf := range confs {
-		confBytes, err := json.Marshal(conf)
-		if err != nil {
-			return nil, err
-		}
-		if len(resultBytes) == 0 {
-			resultBytes = confBytes
-			continue
+	// Sort the confs (potentially) into sets grouped by `mergeValues` fields
+	taggedConfsList, err := handleMergeValues(confs)
+	if err != nil {
+		return nil, err
+	}
+
+	var interfaces []interface{}
+
+	// Merge each tagged sets of confs
+	for _, taggedConfs := range taggedConfsList {
+		confs := taggedConfs.Confs
+
+		resultBytes := []byte{}
+		for i := 0; i < len(confs); i++ {
+			conf := confs[i].Interface()
+			confBytes, err := json.Marshal(conf)
+			if err != nil {
+				return nil, err
+			}
+			if len(resultBytes) == 0 {
+				resultBytes = confBytes
+				continue
+			}
+			resultBytes, err = jsonpatch.MergePatch(resultBytes, confBytes)
+			if err != nil {
+				return nil, err
+			}
 		}
-		resultBytes, err = jsonpatch.MergePatch(resultBytes, confBytes)
+
+		confType := confs[0].Type()
+		result, err := newConf(confType)
 		if err != nil {
 			return nil, err
 		}
-	}
 
-	confType := reflect.TypeOf(confs[0])
-	result, err := newConf(confType)
-	if err != nil {
-		return nil, err
-	}
+		if err := json.Unmarshal(resultBytes, result); err != nil {
+			return nil, err
+		}
```

Comment:
I think this code can be moved to a separate private function:
```go
// mergeJSONPatches merges a list of confs to a single conf using the algorithm described in https://www.rfc-editor.org/rfc/rfc7396
func mergeJSONPatches(confs []interface{}) (interface{}, error) {
```
that's only for readability reasons